### PR TITLE
 Add client adapter for multi-version compatibility 

### LINKED ISSUE #456: Error when server and client are on different version
When the server is on the older version and the client is latest .
Listing , formatting commands fails because of version mismatch 
```
Error: the server could not find the requested resource (get directcsidrives.direct.csi.min.io)
ERROR the server could not find the requested resource (get directcsidrives.direct.csi.min.io)
```

### PR DESCRIPTION
**Pull request by sinhaashish**
Closes #456 

Steps to test 
1. Install `direct-csi` version 1.4.3
```
git checkout v1.4.3
./build.sh
./kubectl-direct_csi --kubeconfig ~/.kube/k3s.yaml install --image direct-csi:v1.4.3
./kubectl-direct_csi --kubeconfig ~/.kube/k3s.yaml  drives list --all -o wide
```
here the server and client bot are on same version so the listing is successful

2. Use the latest client
```
git checkout master
go build ./cmd/kubectl-direct_csi/
./kubectl-direct_csi --kubeconfig ~/.kube/k3s.yaml  drives list --all -o wide
```
This error out stating 
```
Error: the server could not find the requested resource (get directcsidrives.direct.csi.min.io)
ERROR the server could not find the requested resource (get directcsidrives.direct.csi.min.io)
```

3. Use this PR and build the client and then list , format drives , list volumes. It will succeed.

### CODE REVIEW

File: cmd/kubectl-direct_csi/drives_list.go [hunk 1]
```diff
@@ -129,7 +129,8 @@ func listDrives(ctx context.Context, args []string) error {
 
 	filteredDrives, err := getFilteredDriveList(
 		ctx,
-		client.GetDirectCSIClient().DirectCSIDrives(),
+		client.GetUnversionedDirectCSIDriveClientset(),
+		// client.GetDirectCSIClient().DirectCSIDrives(),
```
Comment (wlan0) [inline cmd/kubectl-direct_csi/drives_list.go:133]: remove this comment

File: pkg/client/drive-adapter.go [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	unstructuredObject := finalResult.Object
+	var directCSIDriveList directcsi.DirectCSIDriveList
+	err = runtime.DefaultUnstructuredConverter.FromUnstructured(unstructuredObject, directCSIDriveList)
+	if err != nil {
+		return nil, err
+	}
```
Comment (Praveenrajmani) [inline pkg/client/drive-adapter.go:116]: https://github.com/kubernetes/apimachinery/blob/v0.22.4/pkg/apis/meta/v1/unstructured/unstructured.go#L420

File: pkg/client/drive-adapter.go [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+	finalResult := &unstructured.Unstructured{}
+	err = converter.Migrate(intermediateResult, finalResult, schema.GroupVersion{
+		Version: directcsi.Version,
+		Group:   directcsi.Group,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	unstructuredObject := finalResult.Object
```
Comment (Praveenrajmani) [inline pkg/client/drive-adapter.go:111]: https://github.com/kubernetes/apimachinery/blob/v0.22.4/pkg/apis/meta/v1/unstructured/unstructured.go#L420

**Comment by Praveenrajmani** [comment]
```diff
diff --git a/cmd/kubectl-direct_csi/resource_utils.go b/cmd/kubectl-direct_csi/resource_utils.go
index 4d3b0fb..646073f 100644
--- a/cmd/kubectl-direct_csi/resource_utils.go
+++ b/cmd/kubectl-direct_csi/resource_utils.go
@@ -31,10 +31,10 @@ func getDrivesByIds(ctx context.Context, ids []string) <-chan client.ListDriveRe
        resultCh := make(chan client.ListDriveResult)
        go func() {
                defer close(resultCh)
-               directClient := client.GetDirectCSIClient()
+               directClient := client.GetUnversionedDirectCSIDriveClientset()
                for _, id := range ids {
                        driveName := strings.TrimSpace(id)
-                       d, err := directClient.DirectCSIDrives().Get(ctx, driveName, metav1.GetOptions{})
+                       d, err := directClient.Get(ctx, driveName, metav1.GetOptions{})
                        if err != nil {
                                if !errors.IsNotFound(err) {
                                        klog.ErrorS(err, "could not get drive", driveName)
diff --git a/pkg/client/drive-adapter.go b/pkg/client/drive-adapter.go
index 731d754..c2ef3f5 100644
--- a/pkg/client/drive-adapter.go
+++ b/pkg/client/drive-adapter.go
@@ -18,12 +18,10 @@ package client
 
 import (
        "context"
-       "time"
 
        directcsi "github.com/minio/direct-csi/pkg/apis/direct.csi.min.io/v1beta3"
        "github.com/minio/direct-csi/pkg/converter"
 
-       scheme "github.com/minio/direct-csi/pkg/clientset/scheme"
        clientset "github.com/minio/direct-csi/pkg/clientset/typed/direct.csi.min.io/v1beta3"
        v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 
@@ -32,7 +30,8 @@ import (
        "k8s.io/apimachinery/pkg/runtime/schema"
        types "k8s.io/apimachinery/pkg/types"
        watch "k8s.io/apimachinery/pkg/watch"
-       rest "k8s.io/client-go/rest"
+       "k8s.io/client-go/dynamic"
+       "k8s.io/klog/v2"
 )
 
 const (
@@ -45,12 +44,20 @@ const (
 //var _ rest.Interface = &directCSIDriveAdapter{}
 
 type directCSIDriveAdapter struct {
-       client rest.Interface
+       dynamicClient dynamic.ResourceInterface
        gvk    *schema.GroupVersionKind
 }
 
-func DirectCSIDriveAdapter(r rest.Interface) clientset.DirectCSIDriveInterface {
-       return &directCSIDriveAdapter{client: r}
+func DirectCSIDriveAdapter(d dynamic.Interface) clientset.DirectCSIDriveInterface {
+       // To-Do: Run discovery client to find the api version of the resource
+       dyncamicRc := d.Resource(
+               schema.GroupVersionResource{
+                       Group: "direct.csi.min.io", 
+                       Version: "v1beta2", 
+                       Resource: "directcsidrives",
+               },
+       )
+       return &directCSIDriveAdapter{dynamicClient: dyncamicRc}
 }
 
 // Get takes name of the directCSIDrive, and returns the corresponding directCSIDrive object, and an error if there is any.
@@ -58,13 +65,12 @@ func (c *directCSIDriveAdapter) Get(
        ctx context.Context,
        name string,
        options v1.GetOptions) (*directcsi.DirectCSIDrive, error) {
-       intermediateResult := &unstructured.Unstructured{}
-       err := c.client.Get().
-               Resource("directcsidrives").
-               Name(name).
-               VersionedParams(&options, ParameterCodec).
-               Do(ctx).
-               Into(intermediateResult)
+
+       intermediateResult, err := c.dynamicClient.Get(ctx, name, options, "")
+       if err != nil {
+               klog.Infof("could not get intermediate result: %v", err)
+               return nil, err
+       }
 
        finalResult := &unstructured.Unstructured{}
        err = converter.Migrate(intermediateResult, finalResult, schema.GroupVersion{
@@ -77,136 +83,139 @@ func (c *directCSIDriveAdapter) Get(
 
        unstructuredObject := finalResult.Object
        var directCSIDrive directcsi.DirectCSIDrive
-       err = runtime.DefaultUnstructuredConverter.FromUnstructured(unstructuredObject, directCSIDrive)
+       err = runtime.DefaultUnstructuredConverter.FromUnstructured(unstructuredObject, &directCSIDrive)
        if err != nil {
                return nil, err
        }
        return &directCSIDrive, nil
-
 }

diff --git a/pkg/client/init.go b/pkg/client/init.go
index 64d7f0e..1797047 100644
--- a/pkg/client/init.go
+++ b/pkg/client/init.go
@@ -31,6 +31,7 @@ import (
        "k8s.io/client-go/kubernetes"
        "k8s.io/client-go/metadata"
 
+       "k8s.io/client-go/dynamic"
        "k8s.io/klog/v2"
 )
 
@@ -126,7 +127,12 @@ func Init() {
                klog.Fatalf("could not initialize direct-csi client: %v", err)
        }
 
-       directcsiDriveClientset = DirectCSIDriveAdapter(directCSIClient.RESTClient())
+       dynamicClient, dErr := dynamic.NewForConfig(config)
+       if err != nil {
+               klog.Fatalf("could not initialize dynamic client: %v", dErr)
+               os.Exit(1)
+       }
+       directcsiDriveClientset = DirectCSIDriveAdapter(dynamicClient)
        if err != nil {
                klog.Fatalf("could not initialize direct-csi drive client: %v", err)
        }

```

File: pkg/client/list.go [hunk 4]
```diff
@@ -61,7 +62,8 @@ func ListDrives(ctx context.Context, driveInterface clientset.DirectCSIDriveInte
 			LabelSelector: labelSelector,
 		}
 		for {
-			result, err := driveInterface.List(ctx, options)
+			directClient := GetUnversionedDirectCSIDriveClientset()
```
Comment (Praveenrajmani) [inline pkg/client/list.go:65]: we don't need this, we can make use of the `driveInterface` function argument here.

File: pkg/client/drive-adapter.go [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+	dyncamicRc := d.Resource(
+		schema.GroupVersionResource{
+			Group:    "direct.csi.min.io",
+			Version:  "v1beta2",
+			Resource: "directcsidrives",
+		},
+	)
+	return &directCSIDriveAdapter{dynamicClient: dyncamicRc}
+}
+
```
Comment (Praveenrajmani) [inline pkg/client/drive-adapter.go:84]: We don't need this

**Comment by Praveenrajmani** [comment]
You need to use volume client adapter in `volumes ls` to fix the CI failures.

**Comment by Praveenrajmani** [comment]
Can we also add necessary test cases?

File: pkg/client/client.go [hunk 6]
```diff
@@ -105,7 +104,6 @@ func GetGroupKindVersions(group, kind string, versions ...string) (*schema.Group
 	discoveryClient := GetDiscoveryClient()
 	apiGroupResources, err := restmapper.GetAPIGroupResources(discoveryClient)
 	if err != nil {
-		klog.Errorf("could not obtain API group resources: %v", err)
```
Comment (Praveenrajmani) [inline pkg/client/client.go:108]: can we not remove this log line and set it to some `klog.V(5)` instead?

File: pkg/client/drive.go [hunk 7]
```diff
@@ -156,13 +156,11 @@ func CreateDrive(ctx context.Context, driveInterface clientset.DirectCSIDriveInt
...
 // DeleteDrive deletes drive CRD.
 func DeleteDrive(ctx context.Context, directCSIClient directcsiclientset.Interface, drive *directcsi.DirectCSIDrive, force bool) error {
-	var err error
 	driveInterface := directCSIClient.DirectV1beta3().DirectCSIDrives()
 	volumeInterface := directCSIClient.DirectV1beta3().DirectCSIVolumes()
 	if drive.Status.DriveStatus != directcsi.DriveStatusTerminating {
 		drive.Status.DriveStatus = directcsi.DriveStatusTerminating
-		drive, err = driveInterface.Update(ctx, drive, metav1.UpdateOptions{TypeMeta: DirectCSIDriveTypeMeta()})
-		if err != nil {
+		if _, err := driveInterface.Update(ctx, drive, metav1.UpdateOptions{TypeMeta: utils.DirectCSIDriveTypeMeta()}); err != nil {
```
Comment (Praveenrajmani) [inline pkg/client/drive.go:163]: can we revert this change?

File: pkg/volume/volume_test.go [hunk 8]
```diff
@@ -244,14 +245,15 @@ func TestAbnormalDeleteEventHandle(t *testing.T) {
 
 	vl := createFakeVolumeEventListener(testObjects...)
 	ctx := context.TODO()
-	directCSIClient := vl.directCSIClient.DirectV1beta3()
+	volumeInterface := vl.directCSIClient.DirectV1beta3().DirectCSIVolumes()
```
Comment (Praveenrajmani) [inline pkg/volume/volume_test.go:248]: We don't need this change as we use unversioned clients only in plugin. We can remove the changes in this file.

File: cmd/kubectl-direct_csi/drives_accesstier_set.go [hunk 9]
```diff
@@ -96,10 +96,9 @@ func setAccessTier(ctx context.Context, accessTier directcsi.AccessTier) error {
 	ctx, cancelFunc := context.WithCancel(ctx)
 	defer cancelFunc()
 
-	directCSIClient := client.GetDirectCSIClient()
 	return processFilteredDrives(
 		ctx,
-		directCSIClient.DirectCSIDrives(),
+		client.GetUnversionedDirectCSIDriveClientset(),
```
Comment (wlan0) [inline cmd/kubectl-direct_csi/drives_accesstier_set.go:101]: We don't need to pass this client argument anymore. The variable is initialized in the client package. We can just use it locally in that package. 

We can get rid of the client argument entirely for both processFiltered* and getFiltered*List

File: pkg/client/client.go (hunk 6 above)
Comment (wlan0) [reply pkg/client/client.go:108]: We need this log line. It needs to be klog.Errorf only

File: cmd/kubectl-direct_csi/drives_format_test.go [hunk 10]
```diff
@@ -100,18 +99,18 @@ func TestFormatDrivesByAttributes(t1 *testing.T) {
...
 	defer cancelFunc()
 	testClientSet := clientsetfake.NewSimpleClientset(testDriveObjects...)
-	testClient := testClientSet.DirectV1beta3()
-	client.SetDirectCSIClient(testClient.(*directcsifake.FakeDirectV1beta3))
+	driveInterface := testClientSet.DirectV1beta3().DirectCSIDrives()
+	client.SetUnversionedDirectCSIDriveClientset(driveInterface)
 
 	resetDrives := func() error {
-		driveList, err := client.GetDriveList(ctx, testClient.DirectCSIDrives(), nil, nil, nil)
+		driveList, err := client.GetDriveList(ctx, driveInterface, nil, nil, nil)
```
Comment (wlan0) [inline cmd/kubectl-direct_csi/drives_format_test.go:106]: we dont need to pass the client at all

File: pkg/volume/volume_test.go [hunk 11]
```diff
@@ -244,13 +244,13 @@ func TestAbnormalDeleteEventHandle(t *testing.T) {
...
 	vl := createFakeVolumeEventListener(testObjects...)
 	ctx := context.TODO()
-	directCSIClient := vl.directCSIClient.DirectV1beta3()
 
 	for _, testObj := range testObjects {
 		vObj, ok := testObj.(*directcsi.DirectCSIVolume)
 		if !ok {
 			continue
 		}
+		directCSIClient := vl.directCSIClient.DirectV1beta3()
```
Comment (Praveenrajmani) [inline pkg/volume/volume_test.go:253]: this line has to be outside of the loop, lets revert this change as well

File: pkg/client/fake.go [hunk 12]
```diff
@@ -63,9 +63,12 @@ func FakeInit() {
...
 }
 
-// SetDirectCSIClient sets fake direct-csi client.
-func SetDirectCSIClient(fakeClient *directcsifake.FakeDirectV1beta3) {
-	directCSIClient = fakeClient
+func SetUnversionedDirectCSIDriveClientset(driveInterface directcsiclientset.DirectCSIDriveInterface) {
+	directcsiDriveClientset = driveInterface
+}
+
+func SetUnversionedDirectCSIVolumeClientset(volumeInterface directcsiclientset.DirectCSIVolumeInterface) {
```
Comment (Praveenrajmani) [inline pkg/client/fake.go:70]: can we remove this util function if it is not used anywhere else

File: pkg/client/client.go (hunk 6 above)
Comment (Praveenrajmani) [reply pkg/client/client.go:109]: We call this utility function while initializing adapter clients. And this error is expected for fresh installations (before any CRD resources got created). We don't want this log line printed always for fresh installations.

Comment (wlan0) [reply pkg/client/client.go:109]: ```klog.V(3)```

File: pkg/client/client.go [hunk 13]
```diff
@@ -116,7 +114,6 @@ func GetGroupKindVersions(group, kind string, versions ...string) (*schema.Group
 	}
 	mapper, err := restMapper.RESTMapping(gk, versions...)
 	if err != nil {
-		klog.Errorf("could not find valid restmapping: %v", err)
```
Comment (wlan0) [inline pkg/client/client.go:119]: ```klog.V(3)```

File: pkg/client/volume-adapter.go [hunk 14]
```diff
@@ -0,0 +1,241 @@
...
+	runtime "k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	types "k8s.io/apimachinery/pkg/types"
+	watch "k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/dynamic"
+	rest "k8s.io/client-go/rest"
+	"k8s.io/klog/v2"
+)
+
+var _ clientset.DirectCSIVolumeInterface = &directCSIVolumeAdapter{}
```
Comment (balamurugana) [inline pkg/client/volume-adapter.go:41]: 1. Same as above in drive adapter
2. You could avoid code duplication with above by base directcsi interface and extend that interface and do specifics for drives and volumes.

File: cmd/kubectl-direct_csi/drives_accesstier_set.go [hunk 15]
```diff
@@ -108,7 +106,7 @@ func setAccessTier(ctx context.Context, accessTier directcsi.AccessTier) error {
 			setDriveAccessTier(drive, accessTier)
 			return nil
 		},
-		defaultDriveUpdateFunc(directCSIClient),
+		defaultDriveUpdateFunc(client.GetUnversionedDirectCSIDriveClientset()),
```
Comment (balamurugana) [inline cmd/kubectl-direct_csi/drives_accesstier_set.go:109]: You could name it as `client.GetDirectCSIClientset()` instead of `client.GetUnversionedDirectCSIClientset()`. The specific version of clientset is created by importing specific version so that this name does not conflict.

Comment (sinhaashish) [reply cmd/kubectl-direct_csi/drives_accesstier_set.go]: Renamed it to `GetDirectCSIDriveClientset`

File: pkg/client/volume-adapter.go (hunk 14 above)
Comment (sinhaashish) [reply pkg/client/volume-adapter.go:41]: Second point can be addressed in a different PR. 1st point addressed

File: pkg/client/fake.go (hunk 12 above)
Comment (sinhaashish) [reply pkg/client/fake.go:70]: present to make the code symmetric

File: pkg/client/volume-adapter.go [hunk 16]
```diff
@@ -0,0 +1,241 @@
...
+	}
+	resourceInterface, err := dynamic.NewForConfig(config)
+	if err != nil {
+		return nil, err
+	}
+	dynamicResourceClient := resourceInterface.Resource(
+		schema.GroupVersionResource{
+			Group:    directcsi.Group,
+			Version:  version,
+			Resource: "directcsivolume",
```
Comment (Praveenrajmani) [inline pkg/client/volume-adapter.go:72]:
```suggestion
			Resource: "directcsivolumes",
```

File: functests/common.sh [hunk 17]
```diff
@@ -120,6 +120,7 @@ function deploy_minio() {
 
     required_count=4
     running_count=0
+    echo "hello world"
```
Comment (Praveenrajmani) [inline functests/common.sh]: Can we remove these comments

File: functests/tests.sh [hunk 18]
```diff
@@ -104,6 +104,66 @@ function do_upgrade_test() {
     uninstall_directcsi
 }
 
+function do_version_adapter_test() {
```
Comment (Praveenrajmani) [inline functests/tests.sh:107]: Can we remove the comments in this file

File: functests/tests.sh [hunk 19]
```diff
@@ -104,6 +104,66 @@ function do_upgrade_test() {
...
+    echo "One"
+    wget --quiet --output-document="kubectl-direct_csi_$1" "https://github.com/minio/direct-csi/releases/download/v$1/kubectl-direct_csi_$1_linux_amd64"
+    chmod a+x "kubectl-direct_csi_$1"
```
Comment (Praveenrajmani) [inline functests/tests.sh:109]: We don't need to `wget` again as the `kubectl-direct_csi_1.4.3` and `kubectl-direct_csi_1.3.6` are already downloaded in the previous `upgrade_tests` run.

File: functests/tests.sh [hunk 20]
```diff
@@ -116,3 +164,6 @@ do_upgrade_test "1.3.6"
 
 echo "$ME: Run upgrade test from v1.4.3"
 do_upgrade_test "1.4.3"
+
+echo "$ME: Run version adaptibility test"
+do_version_adapter_test "1.4.3"
```
Comment (Praveenrajmani) [inline functests/tests.sh:168]: Can we rename this func to `do_multi_version_client_test` or some better name?

File: pkg/client/volume-adapter.go [hunk 21]
```diff
@@ -0,0 +1,241 @@
...
+			Group:    directcsi.Group,
+			Version:  version,
+			Resource: "directcsivolumes",
+		},
+	)
+
+	return &directCSIVolumeAdapter{resourceInterface: dynamicResourceClient, gvk: gvk}, nil
+}
+
+// Get takes name of the directCSIVolume, and returns the corresponding directCSVolume object, and an error if there is any.
```
Comment (Praveenrajmani) [inline pkg/client/volume-adapter.go:79]:
```suggestion
// Get takes name of the directCSIVolume, and returns the lates directCSIVolume object, and an error if there is any.
```

**Review by Praveenrajmani — COMMENTED** [review]
We also need to fix `kubectl direct-csi info` command to use drive/volume adapters

File: functests/tests.sh [hunk 22]
```diff
@@ -104,6 +104,50 @@ function do_upgrade_test() {
...
+
+    # Show output for manual debugging.
+    "${DIRECT_CSI_CLIENT}" drives list --all -o wide
+
+    # Show output for manual debugging.
+    "${DIRECT_CSI_CLIENT}" volumes list --all -o wide
+
+    mapfile -t upgraded_volumes < <("${DIRECT_CSI_CLIENT}" volumes list --status published | awk '{print $1}')
+    if [[ ${#upgraded_volumes[@]} -ne ${#volumes[@]} ]]; then
+        echo "$ME: volume count is not matching after upgrade"
```
Comment (Praveenrajmani) [inline functests/tests.sh:135]: can we fix the error message here

```
 volume count is not matching in version compatibility client tests 
```

File: functests/tests.sh [hunk 23]
```diff
@@ -104,6 +104,50 @@ function do_upgrade_test() {
...
+        return 1
+    fi
+
+    for volume in "${upgraded_volumes[@]}"; do
+        if [[ ! ${volumes[${volume}]+_} ]]; then
+            echo "$ME: ${volume} not found after upgrade"
```
Comment (Praveenrajmani) [inline functests/tests.sh:141]: Can we fix the error message here

File: functests/tests.sh [hunk 24]
```diff
@@ -120,3 +164,6 @@ do_upgrade_test "1.3.6"
 
 echo "$ME: Run upgrade test from v1.4.6"
 do_upgrade_test "1.4.6"
+
+echo "$ME: Run version adaptibility test"
+do_version_compatibility_client_test "1.4.3"
```
Comment (Praveenrajmani) [inline functests/tests.sh:169]:
```suggestion
do_version_compatibility_client_test "1.4.6"
```

File: pkg/client/fake.go [hunk 25]
```diff
@@ -63,9 +63,12 @@ func FakeInit() {
 	initEvent(kubeClient)
 }
 
-// SetDirectCSIClient sets fake direct-csi client.
-func SetDirectCSIClient(fakeClient *directcsifake.FakeDirectV1beta3) {
-	directCSIClient = fakeClient
+func SetUnversionedDirectCSIDriveClientset(driveInterface directcsiclientset.DirectCSIDriveInterface) {
```
Comment (Praveenrajmani) [inline pkg/client/fake.go:66]: We can rename these function names to `SetDirectCSIDriveClientset`

**Review by Praveenrajmani — APPROVED** [review]
Unit tests can be added in a separate PR as discussed.

Some minor nits. We can address the in the next PR

**Review by wlan0 — APPROVED** [review]
